### ISSUE TITLE
Add TypeScript types

### ISSUE THREAD

**Issue by mjackson** [ReactTraining/react-router#6955]
We really need to just bring in the TypeScript types and maintain them ourselves. Definitely typed gets behind on our releases, so TypeScript users have to use a lot of `any`s to use our latest stuff.

**Comment by timdorr** [comment]
We recently converted Redux over to TS completely. Unless the plan is to convert the entire source, I would advise against this. 

We had our type defs in the repo, but had issues when the types needed to break compat, but not the library itself. Our choice was to move to DT or convert to TS; we chose the latter.

I don't know your familiarity level with TS, but this would be a fairly large undertaking with such a big repo (at least as far as the mental model goes). It wasn't easy with Redux and we already had a bunch of tooling set up and people helping out. But there were a lot more functional gotchas (`compose` is a giant shit show), and I think this library's types are a lot more straightforward.

I'd just keep that in mind for adding types. It's not really as simple as just throwing together an index.d.ts :)

**Comment by mjackson** [comment]
Thanks for your input, @timdorr. I'm not planning on rewriting in TS anytime soon, but I would like to better support those people who are using the router in a TS project.

Right now I'm thinking we can just start with a basic set of types. They don't need to be exhaustive. The types published on Definitely Typed [are only ~150 lines](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-router/index.d.ts). Seems very straightforward. I guess I need more convincing that this is going to be a huge undertaking, because right now it really doesn't seem like it.

**Comment by lekoaf** [comment]
I just ended up on this issue because DefinitelyTyped does not yet seem to support the new hooks. Which is a shame, because I was gonna start to use them in my typescript project.

**Comment by mjackson** [comment]
@lekoaf I think you may be using an old version. The hooks [are there now](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/05c13bad915e96f1efc6113cabc2bef5e513af1f/types/react-router/index.d.ts#L152-L160).

**Comment by leoyli** [comment]
If the HOC and and other old patterns can start to show deprecation warning in v5.2 or v5.3, I think we can have a better position to migrate to TS in v6.  HOC is a bit hard to be typed, and `compose` is a variadic beast which is unfair to make analogy here (I don't see react-router have this kind of API design). The headache I foresee would be `component` prop (the augmentation pattern), but I think `material-ui` folks have some experiences on that.

**Comment by guyellis** [comment]
@mjackson I agree - I don't think this would be a major undertaking.

In my mind, how I see this happening without converting the codebase to TS and still keeping/checking the types against the code:

- Copy the .d.ts from Definitely Typed and put it in this repo.
- Add `typescript` as a dev dependency
- Add a `tsconfig.json` file that "allowJS" and "noEmit" etc.
- Start adding TypeScript JSDocs in the codebase. (This can be done gradually.)
- Add a call to `tsc` as part of the CI test.

Using JSDocs and a "noop" tsc transpile against the JS code will act as a type linter. So if the types change in the .js code and we don't forget to add/change the JSDocs then `tsc` will catch it.

**Comment by mjackson** [comment]
> if the types change in the .js code and we don't forget to add/change the JSDocs then tsc will catch it.

Is TypeScript able to do type inference on JS code? How will it know that our JSDoc types don't match what's going on in the JS code @guyellis?

**Comment by guyellis** [comment]
> Is TypeScript able to do type inference on JS code?

Yes but it's limited.

> How will it know that our JSDoc types don't match what's going on in the JS code

If we annotate the methods and types in the code with JSDocs then Typescript will catch any invalid usage of those types in the JS code. So adding a new property to an object that hasn't been declared in its interface will cause an error.

Here's a simple repo https://github.com/guyellis/ts-jsdoc

And here are the official docs for TS/JSDocs: https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#supported-jsdoc

Basically we would declare the types in the .d.ts that's provided with the repo when installed and then also consume that same file in JSDocs and use `tsc` in the toolchain to keep everything pukka.

**Comment by xiaoxiangmoe** [comment]
，

**Comment by donaldpipowitch** [comment]
Is there current plan to do it together with V6 or after the V6 release? It could already help to test the alpha releases ❤️ thank you for your work 🙂

**Comment by mjackson** [comment]
The current plan is to convert the v6 codebase to TypeScript. I didn't write it in TypeScript on the initial pass because I just wanted to get it done using the toolset I already knew. But I think I know all the types now. Just have to convert the existing code.

I'd definitely appreciate some help here if someone wants to take a stab at it. All the v6 code is [on the `dev` branch](https://github.com/ReactTraining/react-router/tree/dev).

**Comment by juergenzimmermann** [comment]
@mjackson I just installed `6.0.0-alpha.1`. However, the .d.ts files are not there (yet?).

**Comment by timdorr** [comment]
No conversion has happened yet.

**Comment by ayush987goyal** [comment]
@mjackson Hi! Is there a branch where we can track the TypeScript migration? That way we can help in the migration steps by contributing.

**Comment by timdorr** [comment]
Everything for 6.0 is on the `dev` branch. You can make your contributions there.

**Comment by Therealnaxos** [comment]
I pushed an (attempt) at getting TypeScript support, but would need some help to get it working.

**Comment by sw-yx** [comment]
Therealnaxos' approach involved a source rewrite, but Michael wanted a single .d.ts file. I made a start on it for hacking-around purposes - i have not covered all the api's and i chose to collapse everything into a single dts file including react-router and history, for simplicity in trying stuff out. if you have stuff to add, please comment over there and i will update so taht everyone can benefit 

https://gist.github.com/sw-yx/37a6a3d248c2d4031801f0d568904df8

**Comment by xujin950710** [comment]
Is it possible to use TypeScript for development now, if it is not supported, we ca n’t help with the test

**Comment by murbanowicz** [comment]
Would be great to see that. I hit this issue today as I did want to try v6 in my new project and I cannot do that as there is no types for v6 yet.

**Comment by xiaoxiangmoe** [comment]
We can use [API-Extractor](https://api-extractor.com/) bundle .d.ts files into single file.

**Comment by murbanowicz** [comment]
As I could not stop myself from trying v6 I have created a new version of `@types/react-router` for v6 which I found out worth sharing so others do not have to do hacking with editing types on the fly, `yarn link` it into their project etc.

I made it working nicely with my project, however, I could miss something so please, please, let me know if I should correct anything etc.

What have I done:

- removed obsolete: `Redirect`, `Switch`, `useHistory`
- added: `Routes`, `useNavigation`
- changed: `component/children -> element` in props

PR: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/44015

**Comment by sw-yx** [comment]
oh, good. @murbanowicz i wont be maintaining my gist anymore as i am not actively using react router at the moment. will redirect people to you.

**Comment by mjackson** [comment]
[The codebase was rewritten in TypeScript for v6.0.0-alpha.4](https://github.com/ReactTraining/react-router/releases/tag/v6.0.0-alpha.4), so this is done. Please refer people looking for types to our own built-in types now. Thanks :)
